system/datatypes/mailinglist.php

```diff
@@ -442,6 +452,96 @@ private function unsubscribeAction(HyphaRequest $request) {
 		return ['redirect', isset($_GET['redirect']) ? $_GET['redirect'] : $this->path()];
 	}
 
+	/**
+	 * @param HyphaRequest $request
+	 * @return null|array
+	 */
+	private function remindAction(HyphaRequest $request) {
+		if (!isAdmin()) {
+			notify('error', __(isUser() ? 'admin-rights-needed-to-perform-action' : 'login-to-perform-action'));
+			return ['redirect', $this->path()];
+		}
+
+		$email = isset($_GET['email']) ? $_GET['email'] : null;
+		if (null == $email) {
+			notify('error', __('missing-arguments'));
+			return null;
+		}
+
+		$this->xml->lockAndReload();
+
+		$xpath = './/' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_EMAIL . '=' . xpath_encode($email) . ' and @' . self::FIELD_NAME_STATUS . '!="' . self::ADDRESS_STATUS_UNSUBSCRIBED . '"]';
+		$address = $this->findAddresses($xpath)->first();
+
+		if (!$address instanceof HyphaDomElement) {
+			$this->xml->unlock();
+			notify('error', __('invalid-code'));
+			return null;
+		}
+
+		if ($address->getAttribute(self::FIELD_NAME_STATUS) !== self::ADDRESS_STATUS_PENDING) {
+			$this->xml->unlock();
+			notify('error', __('ml-no-reminder-address-no-longer-pending'));
+			return null;
+		}
+
+		if ($address->getAttribute(self::FIELD_NAME_REMINDED) != true) {
+			$code = $address->getAttribute(self::FIELD_NAME_CONFIRM_CODE);
+			$address->setAttribute(self::FIELD_NAME_REMINDED, true);
+			$this->xml->saveAndUnlock();
+
+			$this->sendConfirmationMail($code, $email);
+			notify('success', ucfirst(__('ml-reminder-sent')));
+		} else {
+			notify('success', ucfirst(__('ml-reminder-already-sent')));
+		}
+
+		return ['redirect', $this->path(self::PATH_ADDRESSES)];
+	}
+
+	/**
+	 * @param HyphaRequest $request
+	 * @return null|array
+	 */
+	private function deleteAddressAction(HyphaRequest $request) {
+		if (!isAdmin()) {
+			notify('error', __(isUser() ? 'admin-rights-needed-to-perform-action' : 'login-to-perform-action'));
+			return ['redirect', $this->path()];
+		}
+
+		$email = isset($_GET['email']) ? $_GET['email'] : null;
+		if (null == $email) {
+			notify('error', __('missing-arguments'));
+			return null;
+		}
+
+		$this->xml->lockAndReload();
+
+		$xpath = './/' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_EMAIL . '=' . xpath_encode($email) . ' and @' . self::FIELD_NAME_STATUS . '!="' . self::ADDRESS_STATUS_UNSUBSCRIBED . '"]';
+		$address = $this->findAddresses($xpath)->first();
+
+		if (!$address instanceof HyphaDomElement) {
+			$this->xml->unlock();
+			notify('error', __('invalid-code'));
+			return null;
+		}
+
+		$status = $address->getAttribute(self::FIELD_NAME_STATUS);
+		if ($status === self::ADDRESS_STATUS_CONFIRMED) {
```

Comment:
Would it not be better to reverse this check? E.g. "only pending addresses can be deleted"? Then this code remains working when additional statuses are added later.

Reply:
true